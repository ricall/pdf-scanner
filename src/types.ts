export type PdfPage = {
    page: number;
    angle: number;
}

export type PdfMetadata = {
    title?: string,
    author?: string,
    subject?: string,
    creator?: string,
    keywords?: string,
    producer?: string,
    creationDate?: Date,
    modificationDate?: Date,
}

export type PdfDocument = {
    name: string,
    bytes: ArrayBuffer,
    pages: PdfPage[],
    currentPage: number,
    metadata: PdfMetadata,
}

export const initialState: PdfDocument = {
    name: 'untitled',
    bytes: new ArrayBuffer(0),
    pages: [],
    currentPage: 0,
    metadata: {
        title: '',
        author: '',
        subject: '',
        creator: '',
        keywords: '',
        producer: '',
        creationDate: undefined,
        modificationDate: undefined,
    }
}

export type PdfAction =
    | { type: 'LOAD', name: string, document: ArrayBuffer, pages: PdfPage[], currentPage: number, metadata: PdfMetadata }
    | { type: 'NEW' }
    | { type: 'SET_CURRENT_PAGE', page: number }
    | { type: 'ROTATE', page: number }
    | { type: 'MOVE_PAGE_LEFT', page: number }
    | { type: 'MOVE_PAGE_RIGHT', page: number }
    | { type: 'DELETE', page: number }
    | { type: 'UPDATE_METADATA', metadata: PdfMetadata }
