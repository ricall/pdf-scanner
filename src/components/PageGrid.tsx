import React from 'react';
import { Grid } from "@material-ui/core";
import Pagination from '@material-ui/lab/Pagination';
import { createStyles, makeStyles, Theme } from "@material-ui/core/styles";
import { Document } from "react-pdf";
import { PdfDocument } from "../types";
import { PageView } from "./PageView";
import { PageAppendPdf } from "./PageAppendPdf";

const useStyles = makeStyles((theme: Theme) => createStyles({
    root: {
        flexGrow: 1,
    },
    paper: {
        width: 270,
        height: 380,
    },
    speedDial: {
        position: 'absolute',
        '&.MuiSpeedDial-directionLeft': {
            bottom: theme.spacing(2),
            right: theme.spacing(2)
        },
    },
    control: {
        padding: theme.spacing(2),
    },
}));

export interface PageGridProps {
    document: PdfDocument;
    setCurrentPage: (page: number) => void;
    onRotate: (page: number) => void;
    onMoveLeft: (page: number) => void;
    onMoveRight: (page: number) => void;
    onDelete: (page: number) => void;
    onAddDocument: (name: string, document: ArrayBuffer) => void;
}

export const PageGrid: React.FC<PageGridProps> = ({
    document,
    setCurrentPage,
    onRotate,
    onMoveLeft,
    onMoveRight,
    onDelete,
    onAddDocument,
    ...props
}) => {
    const classes = useStyles();
    const { bytes, pages, currentPage } = document;

    if (pages.length === 0) {
        return (
            <Grid container className={classes.root} spacing={2}>
                <Grid item xs={12}>
                    <Grid container justifyContent="center" spacing={4}>
                        <Grid item>
                            <PageAppendPdf label="Load PDF" onAddDocument={onAddDocument}/>
                        </Grid>
                    </Grid>
                </Grid>
            </Grid>
        );
    }

    let startPage = Math.max(1, currentPage - 1);
    if (startPage > pages.length - 1) {
        startPage = Math.max(1, pages.length - 1);
    }
    const showDropTarget = startPage >= pages.length - 1;

    return (
        <Document file={bytes}>
            <Grid container className={classes.root} spacing={2}>
                <Grid item xs={12}>
                    <Grid container justifyContent="center" spacing={4}>
                        {[ 0, 1, 2 ]
                            .filter(i => i + startPage <= pages.length)
                            .map(i => {
                                const index = i + startPage;
                                const page = pages[i + startPage - 1];
                                return (
                                    <Grid item key={index}>
                                        <PageView
                                            index={index}
                                            page={page}
                                            selected={currentPage === index}
                                            onClick={setCurrentPage}
                                            onRotatePage={onRotate}
                                            onMovePageLeft={onMoveLeft}
                                            onMovePageRight={onMoveRight}
                                            onDeletePage={onDelete}
                                        />
                                    </Grid>
                                )
                            })}
                        {showDropTarget && (
                            <Grid item>
                                <PageAppendPdf label="Append PDF" onAddDocument={onAddDocument}/>
                            </Grid>
                        )}
                    </Grid>
                </Grid>
                <Grid item xs={12}>
                    <Grid container justifyContent="center" spacing={4}>
                        <Grid item>
                            <Pagination
                                page={currentPage}
                                count={pages.length}
                                showFirstButton
                                showLastButton
                                onChange={(_, p) => setCurrentPage(p)}
                            />
                        </Grid>
                    </Grid>
                </Grid>
            </Grid>
        </Document>
    );
}