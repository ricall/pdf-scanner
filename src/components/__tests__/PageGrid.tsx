import React from 'react';
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { PageGrid } from "../PageGrid";
import { initialState } from "../../types";

describe('PageGrid component', () => {

    it('can be rendered', () => {
        const wrapper = shallow(<PageGrid
            document={initialState}
            setCurrentPage={jest.fn()}
            onRotate={jest.fn()}
            onMoveLeft={jest.fn()}
            onMoveRight={jest.fn()}
            onDelete={jest.fn()}
            onAddDocument={jest.fn()}
        />);
        const tree = toJson(wrapper);
        expect(tree).toMatchSnapshot();
    });

});
