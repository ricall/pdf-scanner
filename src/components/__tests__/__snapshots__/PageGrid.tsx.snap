// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageGrid component can be rendered 1`] = `
<WithStyles(ForwardRef(Grid))
  className="makeStyles-root-1"
  container={true}
  spacing={2}
>
  <WithStyles(ForwardRef(Grid))
    item={true}
    xs={12}
  >
    <WithStyles(ForwardRef(Grid))
      container={true}
      justifyContent="center"
      spacing={4}
    >
      <WithStyles(ForwardRef(Grid))
        item={true}
      >
        <PageAppendPdf
          label="Load PDF"
          onAddDocument={[MockFunction]}
        />
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Grid))>
`;
